lib/Transforms/Scalar/DxilConditionalMem2Reg.cpp

```diff
@@ -110,6 +110,134 @@ class DxilConditionalMem2Reg : public FunctionPass {
     AU.setPreservesCFG();
   }
 
+  // Replace simple array allocas with individual scalar allocas.
+  // Only handle if:
+  // - All the alloca's users are geps
+  // - The geps all have only constant indices
+  // - The geps are indexing to just the scalar elements
+  //
+  bool SplitSimpleAllocas(llvm::Function &F) {
+    llvm::SmallVector<AllocaInst *, 10> ScalarAllocas;
+
+    if (F.empty()) return false;
+    BasicBlock *Entry = &F.getEntryBlock();
+
+    bool Changed = false;
+
+    LLVMContext &Ctx = F.getContext();
+    Module *M = F.getParent();
+    IRBuilder<> Builder(Ctx);
+    DIBuilder DIB(*M);
+    const DataLayout &DL = F.getParent()->getDataLayout();
+
+    for (Instruction *it = &Entry->back(); it != nullptr;) {
```

Comment:
Why start at the back? IIRC, allocas are still all at the front of the block

Reply:
Because we could be deleting GEPs, which could be the next instruction after the alloca. Starting at the back lets me delete instructions on the fly.